tower-batch/src/service.rs

```diff
@@ -2,84 +2,63 @@ use super::{
 {
     tx: mpsc::Sender<Message<Request, T::Future>>,
     handle: Handle,
 }
 
-impl<T, Request> Buffer<T, Request>
+impl<T, Request> Batch<T, Request>
 where
-    T: Service<Request>,
+    T: Service<BatchControl<Request>>,
     T::Error: Into<crate::BoxError>,
 {
-    /// Creates a new `Buffer` wrapping `service`.
+    /// Creates a new `Batch` wrapping `service`.
     ///
-    /// `bound` gives the maximal number of requests that can be queued for the service before
-    /// backpressure is applied to callers.
+    /// The wrapper is responsible for telling the inner service when to flush a
+    /// batch of requests.  Two parameters control this policy:
     ///
-    /// The default Tokio executor is used to run the given service, which means that this method
-    /// must be called while on the Tokio runtime.
+    /// * `max_items` gives the maximum number of items per batch.
+    /// * `max_latency` gives the maximum latency for a batch item.
     ///
-    /// # A note on choosing a `bound`
-    ///
-    /// When `Buffer`'s implementation of `poll_ready` returns `Poll::Ready`, it reserves a
-    /// slot in the channel for the forthcoming `call()`. However, if this call doesn't arrive,
-    /// this reserved slot may be held up for a long time. As a result, it's advisable to set
-    /// `bound` to be at least the maximum number of concurrent requests the `Buffer` will see.
-    /// If you do not, all the slots in the buffer may be held up by futures that have just called
-    /// `poll_ready` but will not issue a `call`, which prevents other senders from issuing new
-    /// requests.
-    pub fn new(service: T, bound: usize) -> Self
+    /// The default Tokio executor is used to run the given service, which means
+    /// that this method must be called while on the Tokio runtime.
+    pub fn new(service: T, max_items: usize, max_latency: std::time::Duration) -> Self
     where
         T: Send + 'static,
         T::Future: Send,
         T::Error: Send + Sync,
         Request: Send + 'static,
     {
-        let (tx, rx) = mpsc::channel(bound);
-        let (handle, worker) = Worker::new(service, rx);
-        tokio::spawn(worker);
-        Buffer { tx, handle }
-    }
-
-    /// Creates a new `Buffer` wrapping `service`, but returns the background worker.
-    ///
-    /// This is useful if you do not want to spawn directly onto the `tokio` runtime
-    /// but instead want to use your own executor. This will return the `Buffer` and
-    /// the background `Worker` that you can then spawn.
-    pub fn pair(service: T, bound: usize) -> (Buffer<T, Request>, Worker<T, Request>)
-    where
-        T: Send + 'static,
-        T::Error: Send + Sync,
-        Request: Send + 'static,
-    {
-        let (tx, rx) = mpsc::channel(bound);
-        let (handle, worker) = Worker::new(service, rx);
-        (Buffer { tx, handle }, worker)
+        // XXX(hdevalence): is this bound good
+        let (tx, rx) = mpsc::channel(1);
```

Comment:
This could cause contention if there are multiple `Batch` handles and one of them calls `poll_ready` but never calls `call`.  However, we use `ready_and` everywhere so I don't think that this is likely to be a problem, and it's appealing to reduce the delay in propagating backpressure.